BibtexQuery/Name.lean

```diff
@@ -0,0 +1,216 @@
+
+- `Lastname, Firstname => (Firstname, Lastname)`
+- `Firstname {Last {N}ame} => (Firstname, Last {N}ame)`
+- `Firstname Lastname => (Firstname, Lastname)`
+- `First Name Lastname => (First Name, Lastname)`
+- `Firstname last Name => (Firstname, last Name)`
+-/
+def getName (arr : Array String) : String × String :=
+  if arr.isEmpty then
+    ("", "")
+  else
+    let join' (arr : Subarray String) : String := arr.foldl (fun acc s =>
+      acc ++ (if acc.isEmpty || s.isEmpty || s == "," then "" else " ") ++ s) ""
+    match arr.getIdx? "," with
+    | .some n =>
+      (arr.toSubarray.drop (n + 1) |> join', arr.toSubarray.take n |> join')
+    | .none =>
+      let s := arr.get! (arr.size - 1)
+      if s.startsWith "{" && s.endsWith "}" then
+        (arr.toSubarray.take (arr.size - 1) |> join',
+          s.toSubstring.drop 1 |>.dropRight 1 |>.toString)
+      else
+        let lastName := arr.reverse.toSubarray.drop 1 |>.toArray.takeWhile
+          (fun s => isLowercase s.front) |>.reverse.push s
+        (arr.toSubarray.take (arr.size - lastName.size) |> join',
+          lastName.toSubarray |> join')
+
+/-- Input a bibtex name string without TeX commands
+and math equations, return an array of `(Firstname, Lastname)`.
+The braces in the name are preserevd. -/
+def getNames : Parsec (Array (String × String)) := do
+  let arr ← getNameAux
+  let arr2 : Array (Array String) := arr.foldl (fun acc s =>
+    if s = "and" then
+      acc.push #[]
+    else
+      acc.modify (acc.size - 1) (Array.push · s)) #[#[]]
+  return arr2.filterMap fun arr =>
+    let ret := getName arr
+    if ret.1.isEmpty && ret.2.isEmpty then .none else .some ret
+
+/-- Strip diacritics from a character. -/
+def stripDiacritics (c : Char) : Char :=
+  match c with
+  | 'œ' => 'o' | 'Œ' => 'O'
+  | 'æ' => 'a' | 'Æ' => 'A'
+  | 'å' => 'a' | 'Å' => 'A'
+  | 'ø' => 'o' | 'Ø' => 'O'
+  | 'ł' => 'l' | 'Ł' => 'L'
+  | 'ı' => 'i'
+  | 'ȷ' => 'j'
+  | '\u00DF' => 's' | '\u1E9E' => 'S'
+  | _ =>
+    let s := getCanonicalDecomposition c
+    s.get? (s.find fun c => (getUnicodeData c).canonicalCombiningClass == 0) |>.getD c
+
+/-- Strip diacritics from a string. -/
+def stripDiacriticsFromString (s : String) : String :=
+  s.toList.toArray.map stripDiacritics |>.filter (not <| GeneralCategory.isMark ·)
+    |>.toList |> String.mk
+
+/-- Check if a string is an upper case Roman numerals.
+It does not check the validity of the number, for example, it accepts `IXIX`. -/
+def isUppercaseRomanNumerals (s : String) : Bool :=
+  not s.isEmpty && s.all fun c =>
+    match c with
+    | 'I' | 'V' | 'X' | 'L' | 'C' | 'D' | 'M' => true
+    | _ => false
+
+/-- Input a last name string without TeX commands, braces
+and math equations, already split by spaces and comma,
+return `(oneLetterAbbr, threeLetterAbbr)` of the last name. -/
+def getLastNameAbbr (arr : Array String) : String × String :=
+  let arr := if arr.size ≤ 1 then arr else arr.filter (not <| isUppercaseRomanNumerals ·)
```

Comment:
This fails to produce a one-letter abbreviation on `#["Dupont-Dupuis"]` (i.e. my full last name) and on `#["Van", "de", "Velde"]`.

Reply:
I'm sorry, but the docstring and the variable names are misleading. It should be remark that one-letter and three-letter abbreviations are not necessary of one-letter and three-letter: they are if there is only one item in the array and there is only one uppercase letter in it. In other cases they are usually not. I think this matches with the behavior of `pybtex`. I'll add a docstring remark this.

Reply:
I have added the docstring. But I'm not sure if the logic matches with the usual behavior of bibtex. Comments welcome.

Reply:
Ah, I see, thanks!